judgels-backends/uriel/uriel-app/src/main/java/judgels/uriel/contest/scoreboard/ScoreboardProcessor.java

```diff
@@ -22,5 +24,22 @@ String computeToString(
             Map<String, Optional<Instant>> contestantStartTimesMap,
             List<Submission> submissions,
             Optional<Instant> freezeTime);
+
+    int getTotalEntries(Scoreboard scoreboard);
+    List<?> getEntries(Scoreboard scoreboard);
+    Scoreboard replaceEntries(Scoreboard scoreboard, List<?> entries);
     Scoreboard filterContestantJids(Scoreboard scoreboard, Set<String> contestantJids);
+
+    default Scoreboard paginate(Scoreboard scoreboard, int page, int pageSize) {
+        List<? extends List<?>> partition = Lists.partition(getEntries(scoreboard), pageSize);
+
+        List partitionPage;
+        try {
+            partitionPage = partition.get(page - 1);
+        } catch (IndexOutOfBoundsException e) {
```

Comment:
This is an antipattern (using exception for normal flow). You can always check if the index is out of bound (using `.size()`)